pandas/core/generic.py

```diff
@@ -6658,9 +6658,8 @@ def replace(
         else:
 
             # need a non-zero len on all axes
-            for a in self._AXIS_ORDERS:
-                if not len(self._get_axis(a)):
-                    return self
+            if not self.size:
```

Comment:
Do we even need this check any more or would this happen implicitly?

Reply:
I'm not sure, I was mostly happy about avoiding the use of AXIS_ORDERS.

Reply:
general question. is NDFrame supposed to represent a n-dimensional data structure or is it a base class for Series and DataFrame?

Reply:
NDFrame is just a base class (of Series & DataFrame)

Reply:
it *was* also backing Panel / Panel4D, but no longer

Reply:
i'm thinking should the generic nature of NDFrame be maintained to facilitate a third party Panel implementation, or is that ruled out-of-scope for pandas?

Reply:
> i'm thinking should the generic nature of NDFrame be maintained to facilitate a third party Panel implementation, or is that ruled out-of-scope for pandas?

no that's out of scope. the reason we removed Panel is because of all of the complexitiy related to > 2ndim. xarray is a better platform for that.

generic is just the collection of common api between Series/DataFrame

Reply:
mypy complained about FrameOrSeries, was OK with NDFrame

Reply:
in pandas._typing FrameOrSeries is a typevar of Series and DataFrame. This is not applicable to core.generic.

in #27646 FrameOrSeries is defined as a typevar bound by NDFrame so that a series returns a series, a DataFrame returns a DataFrame, a subclassed DataFrame returns a subclassed DataFrame etc.

NDFrame is a nominal type, so allows any subclass of NDFrame to be returned.